### DISCUSSION TITLE
Tamagui + Expo + Expo Router + Expo Web

### DISCUSSION THREAD

**Discussion by hirvesh** [tamagui/tamagui#616] [General]
Hello! I'm trying to setup a project which uses Expo, Tamagui and Expo Router and trying to run it off React Native Web.

Having followed the instructions, I am getting the following error when running it on Expo Web:

![CleanShot 2023-01-26 at 17 43 07](https://user-images.githubusercontent.com/67500738/214850483-760e38dd-d208-40c5-afad-f3fd297306d2.png)

Is there anything I am missing?

**Comment by natew** [comment]
Expo Web nor Router aren't documented as one isn't good for web, the other is somewhat beta. But you can modify their webpack config and add the Tamagui webpack setup to see if that works.

Happy to accept PR's to improve docs here too.

**Reply by Albert-Gao** [reply]
The problem with expo router is it uses the Metro for web bundling and I tried the current tamagui metro doc, does not work :D

**Comment by robertherber** [comment]
Would also like this combo to work. I have this feeling Tamagui + Expo Router is the future for Cross Platform :)

**Reply by natew** [reply]
Happy to accept PRs. I’m working on something in this area too but that may be a few months out.

**Comment by robertherber** [comment]
@hirvesh I'm running into this error, have you gotten past it? Tried a few different configurations to no avail.

<img width="741" alt="Screenshot 2023-02-05 at 21 53 17" src="https://user-images.githubusercontent.com/1467411/216845218-df8bbb92-d8b9-486c-b578-0660eccaa764.png">

**Reply by robertherber** [reply]
I did actually get it to build with Expo Router now, with the following patch (using patch-package):
```
diff --git a/node_modules/@tamagui/constants/dist/cjs/constants.js b/node_modules/@tamagui/constants/dist/cjs/constants.js
index dc593df..d67142d 100644
--- a/node_modules/@tamagui/constants/dist/cjs/constants.js
+++ b/node_modules/@tamagui/constants/dist/cjs/constants.js
@@ -35,14 +35,7 @@ const isWeb = process.env.TAMAGUI_TARGET === "web";
 const isWindowDefined = typeof window !== "undefined";
 const isServer = isWeb && !isWindowDefined;
 const isClient = isWeb && isWindowDefined;
-const isRSC = process.env.ENABLE_RSC ? (
-  // note this is statically analyzed so no funny business, just access it without optional chaining
-  // @ts-ignore
-  import_meta.env ? (
-    // @ts-ignore
-    import_meta.env.SSR
-  ) : false
-) : false;
+const isRSC = false;
 const idFn = () => {
 };
 const useIsomorphicLayoutEffect = isRSC ? idFn : isServer ? import_react.useEffect : import_react.useLayoutEffect;
diff --git a/node_modules/@tamagui/constants/dist/esm/constants.js b/node_modules/@tamagui/constants/dist/esm/constants.js
index aef5893..9fb08db 100644
--- a/node_modules/@tamagui/constants/dist/esm/constants.js
+++ b/node_modules/@tamagui/constants/dist/esm/constants.js
@@ -3,14 +3,7 @@ const isWeb = process.env.TAMAGUI_TARGET === "web";
 const isWindowDefined = typeof window !== "undefined";
 const isServer = isWeb && !isWindowDefined;
 const isClient = isWeb && isWindowDefined;
-const isRSC = process.env.ENABLE_RSC ? (
-  // note this is statically analyzed so no funny business, just access it without optional chaining
-  // @ts-ignore
-  import.meta.env ? (
-    // @ts-ignore
-    import.meta.env.SSR
-  ) : false
-) : false;
+const isRSC = false;
 const idFn = () => {
 };
 const useIsomorphicLayoutEffect = isRSC ? idFn : isServer ? useEffect : useLayoutEffect;
diff --git a/node_modules/@tamagui/constants/src/constants.ts b/node_modules/@tamagui/constants/src/constants.ts
index da020d4..366cdf1 100644
--- a/node_modules/@tamagui/constants/src/constants.ts
+++ b/node_modules/@tamagui/constants/src/constants.ts
@@ -6,14 +6,7 @@ export const isServer = isWeb && !isWindowDefined
 export const isClient = isWeb && isWindowDefined
 
 // may want to move to VITE_RSC_BUILD
-export const isRSC = process.env.ENABLE_RSC
-  ? // note this is statically analyzed so no funny business, just access it without optional chaining
-    // @ts-ignore
-    import.meta.env
-    ? // @ts-ignore
-      import.meta.env.SSR
-    : false
-  : false
+export const isRSC = false
 
 const idFn = () => {}
 export const useIsomorphicLayoutEffect = isRSC
```

Don't really have enough context to make a PR of this at this stage, but essentially removing the `import.meta` ESM stuff seems enough to get it running with Metro / Expo Router. Maybe it would make sense to just replace the `import.meta` references in the source files @natew?

My babel.config.js looks like this:
```JavaScript
module.exports = function (api) {
  api.cache(true)

  return {
    presets: ['babel-preset-expo'],
    plugins: [
      // NOTE: `expo-router/babel` is a temporary extension to `babel-preset-expo`.
      require.resolve('expo-router/babel'),
      [
        'transform-inline-environment-variables', {
          include: 'TAMAGUI_TARGET',
        },
      ],
    ],
  }
}
```

And script to start it all:
`"web": "TAMAGUI_TARGET=web expo start --web --clear",`

**Comment by matthewmturner** [comment]
could `npm create tamagui` be updated to have the expo app use the latest expo sdk (48) and have it use expo-router?  To be clear, not trying to use expo-router for web navigation but would prefer to use it on native now that v1 release is out and its my understanding that [solito also can be used with expo router](https://solito.dev/expo-router).

**Comment by G-Ray** [comment]
I created a starter with expo + expo-router + tamagui to develop universal apps : 
https://github.com/G-Ray/expo-tamagui-universal-starter

I applied the same patch to `@tamagui/constants` as @robertherber (in esm build only) using `patch-package`.

Tamagui's configuration replicates what is in the [guide](https://tamagui.dev/docs/guides/expo), although I added web support for `react-native-reanimated` with `@babel/plugin-proposal-export-namespace-from` as described in the [expo guide](https://docs.expo.dev/versions/latest/sdk/reanimated/#web-support).

Feel free to test it and contribute :rocket:

**Comment by localjo** [comment]
I've been trying to get Tamagui working with Expo Web. According to the [Expo guide](https://tamagui.dev/docs/guides/expo), "it should work with minimal effort as Tamagui doesn't require any extra configuration beyond aliasing react-native to react-native-web", but I am not finding that to be the case. I tried the steps in the guide, and @robertherber 's patch file, and @G-Ray 's starter. Unfortunately I'm having a hard time getting it to work.

I'm seeing the same error as @hirvesh in the browser. And in the terminal, I'm seeing:

```
./node_modules/@radix-ui/react-use-escape-keydown/dist/index.mjs 9:28-49
Can't import the named export 'useCallbackRef' from non EcmaScript module (only default export is available)
./node_modules/@radix-ui/react-use-escape-keydown/dist/index.mjs 10:4-20
Can't import the named export 'useEffect' from non EcmaScript module (only default export is available)
./node_modules/react-native-gesture-handler/lib/module/components/GestureComponents.web.js:1
Module not found: Can't resolve 'react-native-web/dist/exports/DrawerLayoutAndroid'
> 1 | import * as React from 'react';
  2 | import {
  3 |   DrawerLayoutAndroid as RNDrawerLayoutAndroid,
  4 |   FlatList as RNFlatList,
./node_modules/@tamagui/radio-group/dist/esm/RadioGroup.js 100:66
Module parse failed: Unexpected token (100:66)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|       "...size": (value, { props }) => {
|         const size = Math.floor(
>           getVariableValue(getSize(value)) * (props["scaleSize"] ?? 0.5)
|         );
|         return {
./node_modules/@tamagui/web/dist/esm/createComponent.js 106:49
Module parse failed: Unexpected token (106:49)
You may need an appropriate loader to handle this file type...
```

I haven't been able to digest the meaning of this yet, but it suggests to me that there is some kind of problem with the way that Tamagui is transpiling the code for the web target, or that there is a need for another plugin or setting in my babel.config.js. I'm using the Babel config suggested in the Expo guide, and I get the same error with and without the optional `@tamagui/babel-plugin` included.

**Comment by okwast** [comment]
Hey everyone,

Has anyone tried that with Expo 49 and Expo Router v2 yet?
It is running perfectly fine on mobile following the expo setup steps but on web I get some weird errors I couldn't figure out yet.

* `Unable to resolve "@radix-ui/react-use-previous" from "node_modules/@tamagui/checkbox/dist/esm/Checkbox.js"`
* `Metro error: assets.filter is not a function`

Seems to irrelevant if I set "native" or "web" in the environment.
